src/Common/FST.h

```diff
@@ -0,0 +1,171 @@
+#pragma once
+#include <array>
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include <Core/Types.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteBuffer.h>
+#include <base/types.h>
+
+namespace DB
+{
+/// Finite State Transducer is an efficient way to represent term dictionary.
+/// It can be viewed as a map of <term, offset>.
```

Comment:
Since FST is considered a generic compact data structure, I corrected it by changing `offset` to `output` in the comment. In the meanwhile, more comments put in the place where FST is actually used, which is  offsets in `struct TermDictionary`.